Updates/fixes for llama.cpp textgen settings

DocShotgun opened the pull request:
## Several changes to the llama.cpp textgen settings:

* Add the llama.cpp-exclusive [`min_keep`](https://github.com/ggml-org/llama.cpp/blob/15a28ec8c705b188ebe178170966d1dcc36fe151/common/common.h#L133) parameter that constrains the effect of truncation samplers.
* Add penalties (rep pen, freq pen, and pres pen) to the llama.cpp sampler order block, allowing them to actually be applied.
* Add support for `nsigma` for llama.cpp (https://github.com/ggml-org/llama.cpp/pull/13264)
    - Addresses https://github.com/SillyTavern/SillyTavern/issues/3960
    - Add `nsigma` to the sampler order block.
    - Add the llama.cpp alias `top_n_sigma` for `nsigma`.
    - Add the ability to set `nsigma` to a negative value, as this represents "disabled" in llama.cpp (https://github.com/ggml-org/llama.cpp/pull/13264/commits/9152659426e2c6d8141d11d447ae9a1463b76f7f). 0 behaves as deterministic in llama.cpp. I chose the negative value -0.01 instead of -1 to avoid extending the slider too far into negative territory, but yeah it's ugly lol. Not sure if there is a better solution.
* Remove [`tfs`](https://github.com/ggml-org/llama.cpp/pull/10071) and `top_a` for llama.cpp as these samplers are not supported.
    - Remove the llama.cpp-specific alias for `tfs`, `tfs_z`.
* Correct the identification string for `typical_p` to the canonical `typ_p` in the llama.cpp sampler order block, allowing it to actually be applied. The [valid options](https://github.com/ggml-org/llama.cpp/blob/a7366faa5bb2fff97b9fb43340d853709f52d8c9/common/sampling.cpp#L501-L528) are `typ_p`, `typical-p`, `typical`, `typ-p`, and `typ`, but not... `typical_p`.


## Checklist:

- [X] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

DocShotgun committed:
Update llama.cpp textgen settings
* Add min_keep, a llama.cpp-exclusive setting for constraining the effect of truncation samplers
* Enable nsigma for llama.cpp, and add llama.cpp alias top_n_sigma, add nsigma to the llama.cpp sampler order block
* Allow a negative value of nsigma as this represents 'disabled' in llama.cpp (while 0 is deterministic)
* Remove tfs and top_a as these are not supported by llama.cpp (tfs was removed, and top_a was never supported)
* Correct the identification string for typical_p in the llama.cpp sampler order block
* Add penalties to the llama.cpp sampler order block

DocShotgun committed:
Remove tfs_z alias for llama.cpp
* This sampler is no longer supported in llama.cpp

Beinsezii commented:
>  Add the ability to set `nsigma` to a negative value, as this represents "disabled" in llama.cpp (https://github.com/ggml-org/llama.cpp/pull/13264/commits/9152659426e2c6d8141d11d447ae9a1463b76f7f). 0 behaves as deterministic in llama.cpp. I chose the negative value -0.01 instead of -1 to avoid extending the slider too far into negative territory, but yeah it's ugly lol. Not sure if there is a better solution.

I'm wondering if it's worth petitioning upstream to change the check to <= 0? They recently changed the 0 behavior of top_k at my suggestion from a full logit sort to a noop. I haven't read the n_sigma paper yet so I'm not sure if that loses any useful function.

Otherwise you could also exclude n_sigma from the sampler order sent by ST entirely if it's == 0. Uglier code but prettier UI maybe. Could extend to other samplers too by defining a "disabled" associated const. Based on the top_k convo that's their intended use case, and the short circuit is more by demand.

DocShotgun committed:
Translate nsigma 0 to -1 to disable for llama.cpp

DocShotgun commented:
> I'm wondering if it's worth petitioning upstream to change the check to <= 0? They recently changed the 0 behavior of top_k at my suggestion from a full logit sort to a noop. I haven't read the n_sigma paper yet so I'm not sure if that loses any useful function.

AFAICT, `n_sigma` = 0 behaves the same as `top_k` = 1 or `temperature` = 0, where only the top candidate is kept. I don't think there's much utility to this as those options already exist.

https://github.com/ggml-org/llama.cpp/pull/13345 ~~Here, I went ahead and proposed the change, we'll see what they say.~~

> Otherwise you could also exclude n_sigma from the sampler order sent by ST entirely if it's == 0. Uglier code but prettier UI maybe. Could extend to other samplers too by defining a "disabled" associated const. Based on the top_k convo that's their intended use case, and the short circuit is more by demand.

~~For now, I've done the simpler code change which is just to translate 0 -> -1 specifically for the llama.cpp backend.~~

Upstream check now changed to <= 0, so I've reverted the special case handling for llama.cpp.

DocShotgun committed:
Remove special handling of nsigma for llama.cpp
* 0 now changed to disable/no-op upstream
